tests/integration/clientv3/watch/cache_test.go

```diff
@@ -0,0 +1,329 @@
+		t.Run(tt.name, func(t *testing.T) {
+			c, err := cache.New(
+				client, prefix,
+				cache.WithHistoryWindowSize(tt.window),
+				cache.WithPerWatcherBufferSize(0),
+				cache.WithResyncInterval(10*time.Millisecond),
+			)
+			if err != nil {
+				t.Fatalf("New(...): %v", err)
+			}
+			defer c.Close()
+
+			if err := c.WaitReady(t.Context()); err != nil {
+				t.Fatalf("cache not ready: %v", err)
+			}
+			ch := c.Watch(t.Context(), prefix, clientv3.WithPrefix())
+
+			generateEvents(t, client, prefix, tt.eventCount)
+			gotEvents, ok := collectAndAssertAtomicEvents(t, ch)
+			closed := !ok
+
+			if tt.wantExactEventCount != 0 && tt.wantExactEventCount != len(gotEvents) {
+				t.Errorf("gotEvents=%v, wantEvents=%v", len(gotEvents), tt.wantExactEventCount)
+			}
+			if tt.wantAtMaxEventCount != 0 && len(gotEvents) > tt.wantAtMaxEventCount {
+				t.Errorf("gotEvents=%v, wantEvents<%v", len(gotEvents), tt.wantAtMaxEventCount)
+			}
+			if closed != tt.wantClosed {
+				t.Errorf("closed=%v, wantClosed=%v", closed, tt.wantClosed)
+			}
+		})
+	}
+}
+
+func TestCacheUnsupportedWatchOptions(t *testing.T) {
+	integration.BeforeTest(t)
+	clus := integration.NewCluster(t, &integration.ClusterConfig{Size: 1})
+	t.Cleanup(func() { clus.Terminate(t) })
+	client := clus.Client(0)
+
+	c, err := cache.New(client, "", cache.WithHistoryWindowSize(1))
+	if err != nil {
+		t.Fatalf("cache.New: %v", err)
+	}
+	defer c.Close()
+	if err := c.WaitReady(t.Context()); err != nil {
+		t.Fatalf("cache not ready: %v", err)
+	}
+
+	unsupported := []struct {
+		name string
+		opt  clientv3.OpOption
+	}{
+		{"WithPrevKV", clientv3.WithPrevKV()},
+		{"WithFragment", clientv3.WithFragment()},
+		{"WithProgressNotify", clientv3.WithProgressNotify()},
+		{"WithCreatedNotify", clientv3.WithCreatedNotify()},
+		{"WithFilterPut", clientv3.WithFilterPut()},
+		{"WithFilterDelete", clientv3.WithFilterDelete()},
+	}
+
+	for _, tc := range unsupported {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			ch := c.Watch(t.Context(), "foo", tc.opt)
+
+			resp, ok := <-ch
+			if !ok {
+				t.Fatalf("channel closed without yielding a response")
+			}
+			if !resp.Canceled {
+				t.Errorf("expected Canceled=true, got %+v", resp)
+			}
+			if !strings.Contains(resp.Err().Error(), cache.ErrUnsupportedRequest.Error()) {
+				t.Errorf("expected ErrUnsupportedWatch text %q, got %v",
+					cache.ErrUnsupportedRequest.Error(), resp.Err())
+			}
+		})
+	}
+}
+
+func generateEvents(t *testing.T, client *clientv3.Client, prefix string, n int) {
+	t.Helper()
+	for i := 0; i < n; i++ {
+		key := fmt.Sprintf("%s%d", prefix, i)
+		if _, err := client.Put(t.Context(), key, fmt.Sprintf("%d", i)); err != nil {
+			t.Fatalf("Put(%q): %v", key, err)
+		}
+	}
+}
+
+type Watcher interface {
+	Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan
+}
+
+type Getter interface {
+	Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error)
+}
```

Comment:
Curious why not to add the interfaces into cache.go as the formal interfaces that the cache supports?

Also I see that they are just part of the clientv3's interface, does it mean that the cache only support a subset interface of clientv3?

Reply:
> Curious why not to add the interfaces into cache.go as the formal interfaces that the cache supports?

Those interfaces are just for testing as the TestWatch only needs those two methods.

> Also I see that they are just part of the clientv3's interface, does it mean that the cache only support a subset interface of clientv3?

Yes, you cannot cache write methods.